docker/dockerfiles/Dockerfile

```diff
@@ -1,154 +1,215 @@
+# compile node to /node, this layer is merged at the end
+#########################################################
 
-# more tools
 RUN apt-get update && \
-    apt-get install -y \
-      git \
-      xvfb \
-      yarn=1.22.5-1 \
-      sudo \
-      openssh-client \
-      ca-certificates \
-      tar \
-      gzip \
-      wget \
-      xz-utils \
-      autoconf \
-      build-essential \
-      zlib1g-dev \
-      libssl-dev \
-      curl \
-      libreadline-dev \
-      python python-dev \
-      vim
+    apt-get install -y --no-install-recommends \
+    build-essential \
+    curl \
+    g++ \
+    python3-pip
 
 # install node
-RUN wget https://nodejs.org/dist/v14.17.1/node-v14.17.1.tar.gz && \
-    tar -xzvf node-v14.17.1.tar.gz && \
+RUN curl -sSLOJ https://nodejs.org/dist/v14.17.1/node-v14.17.1.tar.gz && \
+    tar -xzf node-v14.17.1.tar.gz && \
     rm node-v14.17.1.tar.gz && \
     cd node-v14.17.1 && \
     ./configure && \
-    make -j4 && \
-    make install && \
+    make -j `nproc` && \
+    make install PREFIX=/build && \
     cd .. && \
     rm -r node-v14.17.1
 
-# more more tools
-RUN apt-get install -y unzip
+
+#########################################################
+FROM base as rbenv
+#########################################################
+# compile ruby to /home/${USER}/.rbenv
+#########################################################
+
+USER ${USER}
+
+RUN mkdir -p "$(rbenv root)"/plugins && \
+    git clone https://github.com/rbenv/ruby-build.git "$(rbenv root)"/plugins/ruby-build && \
+    rbenv install 2.7.5 && \
+    eval "$(rbenv init -)" && \
+    rbenv shell 2.7.5 && \
+    rbenv rehash && \
+    gem install bundler -v 2.3.22
+
+
+#########################################################
+FROM base as primary-layer
+#########################################################
+# Main layer of this Dockerfile: add packages etc below
+#########################################################
+
+# MAINLINE UBUNTU PACKAGE DEPS SHOULD BE ADDED HERE
+RUN apt-get install -y --no-install-recommends \
+    apt-transport-https \
+    ca-certificates \
+    enscript \
+    ghostscript \
+    gnupg \
+    gzip \
+    imagemagick \
+    jq \
+    libaio-dev \
+    libaio1 \
+    libicu-dev \
+    libmagickwand-dev \
+    libsqlite3-dev \
+    locales \
+    moreutils \
+    openssh-client \
+    parallel \
+    python \
+    python-dev \
+    software-properties-common \
+    sudo \
+    tar \
+    tzdata \
+    unzip \
+    vim \
+    xvfb \
+    xz-utils \
+    zlib1g-dev \
+    && true
+
+WORKDIR /tmp
+
+# en_US.UTF-8 locale not available by default
+RUN locale-gen en_US.UTF-8
+
+# install mysql
+RUN curl -sSLOJ https://dev.mysql.com/get/mysql-apt-config_0.8.22-1_all.deb && \
+    echo "mysql-apt-config mysql-apt-config/select-server select mysql-5.7" | /usr/bin/debconf-set-selections && \
+    dpkg -i mysql-apt-config_0.8.22-1_all.deb || apt-get -fy install && \
```

Comment:
are the `DEBIAN_FRONTEND=noninteractive` directives no longer necessary?

Reply:
I set them as an ENV at the top, so they apply to all apt-get commands rather than being remembered in some and forgotten in others. If you miss one, all the debconf that's backlogged can drop on your head, so its better to go all-in on noninteractive to prevent future surprises.

```
ENV \
    LANG=C.UTF-8 \
    TZ=Etc/UTC \
    DEBIAN_FRONTEND=noninteractive \
    USER=circleci \
    GROUP=circleci
```

Reply:
Oh, very nice! Yeah, much prefer this being set globally. Thanks!